### PR TITLE
Enforce content-type on client and server.

### LINKED ISSUE #360: Check response content type on client
[It seems](https://github.com/grpc/grpc-java/blob/d0aad72441f2f110910003599bb923db6f671f02/core/src/main/java/io/grpc/transport/Http2ClientStream.java#L49) we aren't checking the content type of responses on the client. The flag was originally put in place because some implementations were sending the wrong content type, but it seems we should enable it again.

### PR DESCRIPTION
**Pull request by nmittler**
Fixes #360

### CODE REVIEW

**Comment by nmittler** [comment]
@ejona86 PTAL

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java [hunk 1]
```diff
@@ -87,24 +86,34 @@ protected void transportHeadersReceived(Metadata headers) {
...
-    } else {
-      stripTransportDetails(headers);
-      inboundHeadersReceived(headers);
+      return;
     }
+
+    // Verify that the content type is appropriate for gRPC.
+    transportError = checkContentType(headers);
+    if (transportError != null) {
+      // The request contained an invalid Content-Type, cancel the stream now.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: Why cancel the stream now and not accumulate the body's content like was already coded?

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java [hunk 2]
```diff
@@ -205,9 +214,9 @@ private Status checkContentType(Metadata headers) {
     }
     contentTypeChecked = true;
     String contentType = headers.get(GrpcUtil.CONTENT_TYPE_KEY);
-    if (TEMP_CHECK_CONTENT_TYPE && !GrpcUtil.CONTENT_TYPE_GRPC.equalsIgnoreCase(contentType)) {
-      // Malformed content-type so report an error
-      return Status.INTERNAL.withDescription("invalid content-type " + contentType);
+    if (contentType != null && !GrpcUtil.isGrpcContentType(contentType)) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: No need to check `contentType != null`

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: The thought here is that if we're receiving a successful response with an incorrect content-type, we have no idea how much data might follow.  Probably best to just terminate the stream now.

This is different than if we receive an HTTP response with an error. In that case, it's likely that whatever comes afterward will be error/debug information that might be useful.

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java (hunk 2 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: Ok, so we require that the first received HEADERS frame contains the content-type? Makes sense.

Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: done.

**Comment by nmittler** [comment]
@ejona86  PTAL

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java (hunk 1 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: > we have no idea how much data might follow

The current behavior already handles that:

```java
if (transportError.getDescription().length() > 1000 || endOfStream) {
```

> Probably best to just terminate the stream now.

Except that data is really useful in figuring out what went wrong. You _need_ it for debugging, and providing it in the status can making finding a bug 5 seconds long instead of 5 days as you try to figure out logging and reproducing in production.

> This is different than if we receive an HTTP response with an error. In that case, it's likely that whatever comes afterward will be error/debug information that might be useful.

Sorta. But let's say that a page existed, instead of a gRPC service. It is totally believable that it may respond 200 but not provide an error, because it isn't talking gRPC and there wasn't any error as far as it was concerned. The information that came afterward would likely be an HTML page and could easily provide more context as to what went wrong.

**Comment by nmittler** [comment]
@ejona86 PTAL

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: done.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java [hunk 3]
```diff
@@ -199,11 +203,12 @@ protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
       StreamException http2Ex) {
     logger.log(Level.WARNING, "Stream Error", cause);
     Http2Stream stream = connection().stream(Http2Exception.streamId(http2Ex));
-    if (stream != null) {
+    NettyServerStream serverStream = stream != null ? serverStream(stream) : null;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: Was this change necessary for verifying content type?

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java [hunk 4]
```diff
@@ -304,6 +309,19 @@ private void cancelStream(ChannelHandlerContext ctx, CancelServerStreamCommand c
     encoder().writeRstStream(ctx, cmd.stream().id(), Http2Error.CANCEL.code(), promise);
   }
 
+  private void verifyContentType(int streamId, Http2Headers headers) throws Http2Exception {
+    ByteString contentType = headers.get(CONTENT_TYPE_HEADER);
+    if (contentType == null) {
+      throw Http2Exception.streamError(streamId, Http2Error.REFUSED_STREAM,
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: We should respond with a clean status (via headers). We _must_ not use REFUSED_STREAM though, as that would imply that the client should re-issue the request.

> The REFUSED_STREAM error code can be included in a RST_STREAM frame to indicate that the stream is being closed prior to any processing having occurred. Any request that was sent on the reset stream can be safely retried.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java [hunk 5]
```diff
@@ -304,6 +309,19 @@ private void cancelStream(ChannelHandlerContext ctx, CancelServerStreamCommand c
...
+          "Content-Type is missing from the request");
+    }
+    String contentTypeString = contentType.toString(UTF_8);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: The value should be ASCII.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java (hunk 3 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: It actually wasn't necessary ... I was verifying the content-type too early, before the stream was set.  Reverted.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java (hunk 5 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: done.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java (hunk 4 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: Currently, this code is consistent with `determineMethod` ... if this needs to change, I think they both do.  WDYT? What should be the error? INTERNAL?

**Comment by nmittler** [comment]
@ejona86 PTAL

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java [hunk 6]
```diff
@@ -205,9 +204,9 @@ private Status checkContentType(Metadata headers) {
     }
     contentTypeChecked = true;
     String contentType = headers.get(GrpcUtil.CONTENT_TYPE_KEY);
-    if (TEMP_CHECK_CONTENT_TYPE && !GrpcUtil.CONTENT_TYPE_GRPC.equalsIgnoreCase(contentType)) {
-      // Malformed content-type so report an error
-      return Status.INTERNAL.withDescription("invalid content-type " + contentType);
+    if (!GrpcUtil.isGrpcContentType(contentType)) {
+      return Status.INTERNAL.withDescription("Invalid content-type: " + contentType + ".\n"
+              + headers.toString());
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: Don't include headers. They will be included later.

File: netty/src/main/java/io/grpc/netty/NettyServerHandler.java (hunk 4 above)
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServerHandler.java]: As discussed offline, yeah INTERNAL seems fine for now. In the future we'll want to use response headers to communicate a status.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 7]
```diff
@@ -397,7 +416,7 @@ private ByteBuf badFrame() throws Exception {
     return captureWrite(ctx);
   }
 
-  private static NettyServerHandler newHandler(ServerTransportListener transportListener,
+  private NettyServerHandler newHandler(ServerTransportListener transportListener,
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: Why the removal of the static keywords?

File: core/src/main/java/io/grpc/internal/Http2ClientStream.java (hunk 6 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/internal/Http2ClientStream.java]: done.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 7 above)
Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: no idea ;) ... reverted.

**Comment by nmittler** [comment]
@ejona86 PTAL

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 8]
```diff
@@ -406,7 +425,7 @@ private static NettyServerHandler newHandler(ServerTransportListener transportLi
         Integer.MAX_VALUE, flowControlWindow);
   }
 
-  private static NettyServerHandler newHandler(ServerTransportListener transportListener) {
+  private NettyServerHandler newHandler(ServerTransportListener transportListener) {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: Ditto static removal

Comment (nmittler) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: done.

**Comment by nmittler** [comment]
@ejona86 anything else or shall I cherry-pick?

**Comment by carl-mastrangelo** [comment]
@nmittler LGTM

**Comment by ejona86** [comment]
@nmittler LGTM

**Comment by nmittler** [comment]
Cherry-picked as d3ab427
